tbot_contrib/locking.py

```diff
@@ -0,0 +1,166 @@
+
+
+class MachineLock(machine.PreConnectInitializer):
+    """
+    Machine locking implementation based on Python and bash.
+    """
+    @property
+    def lock_name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def lock_expiry(self) -> int:
+        return -1
+
+    @property
+    def lock_pool_max(self) -> int:
+        return 1
+
+    @property
+    def lock_sleep(self) -> int:
+        return 1
+
+    lock_no = -1
+    lock_file = ""
+
+    @contextlib.contextmanager 
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as lh:
+            result = lh.exec("mktemp", "lock.XXX")
+            if result[0] == 0:
+                tempfile = result[1].strip()
+                if self.lock_expiry < 1:
+#                    lh.exec("echo", "0", linux.RedirStdout(tempfile))
+                    lh.exec("echo", "0", linux.Raw(">"), tempfile)
+                else: # Write to file the Unix time timestamp of expiry
+#                    lh.exec("echo", str(int(time.time())+self.lock_expiry), linux.RedirStdout(tempfile))
+                    lh.exec("echo", str(int(time.time())+self.lock_expiry), linux.Raw(">"), tempfile)
+            else:
+                raise Exception("Could not create tempfile")
+            repeat = True
+            while (repeat == True):
+                repeat = False
+                lock_no = -1
+                for count in range(self.lock_pool_max): # Iterate through locks
+                    self.lock_file = f"{self.lock_name}{count}"
+                    print(f"Try locking on {count} using {self.lock_file}...")
+                    result = lh.exec("ln", tempfile, self.lock_file) # Is atomic 
+                    if result[0] == 0: # We got a lock, we're done
+                        repeat = False; 
+                        self.lock_no = count
+                        break
+                    else: # No lock this time
+                        # Check if the lock is stale. This may race!
+                        result = lh.exec("cat", self.lock_file)
+                        if result[0] == 0:
+                            if (int(result[1]) != 0) and (int(result[1]) < int(time.time())):
+                                lh.exec("rm", self.lock_file) # Delete stale lock
+                                print(f"Lock {self.lock_file} expired, deleted.")
+                                repeat = True; # The stale lock may be available next time round.
+                if self.lock_no != -1: # If we got a lock, check that is still ours
+                    time.sleep(self.lock_sleep) # Allow possible multiple expired lock deletes to settle
+                    result = lh.exec("stat", "-c%i", self.lock_file)
+                    if result[0] == 0: # lock still exists
+                        if result[1] != lh.exec("stat", "-c%i", tempfile)[1]:
+                            repeat = True; # but inodes don't match, try again
```

Comment:
Decompose directly instead of then having to deal with a tuple and indexing:

```python
retcode, output = lh.exec("stat", "-c%i", self.lock_file)
if retcode == 0: # lock still exists
    if output != lh.exec("stat", "-c%i", tempfile)[1]:
        repeat = True; # but inodes don't match, try again
```